README.md
## nexus-ilegacy
Tool to recursively import maven, npm and nuget libraries to a repository like Nexus OSS, Artifactory, NPM and so on.

---

* [About](#about)
* [Prerequisites](#prerequisites)
* [Installing](#installing)
* [Usage](#usage)
* [History](#history)
* [License](#license)

---

## About
The purpose of this CLI is to import Maven, NPM or NuGet artifacts recursively to the Nexus OSS. It all began when we started to migrate from the Artifactory repo to the Nexus OSS repo at work. At that moment I didn't find a tool or functionality at Nexus to get the job done. So to avoid repetitive work I decided to write a simple cli that calls mvn, npm and nuget behind, taking a directory as starting point, to **recursively** search for artifacts to deploy to the Nexus OSS 3x.

During the execution is possible to check the total of import success and errors. In the end the CLI will generate a log file to better evaluate the results.

## Prerequisites

The nexus-ilegacy CLI requires:
* Node 6.9.0 or higher
* NPM 3 or higher
* Maven 3.3 or higher - to import Maven artifacts
* NuGet 3 or higher - to import .NET artifacts

## Installing

To install the CLI:
```bash
npm install i -g nexus-ilegacy
```

Docs install the Maven and NuGet
* [Maven](https://maven.apache.org/install.html)
* [NuGet](https://www.microsoft.com/net/core#linuxredhat)

**Important**:
1. Both Maven and NuGet need to be on OS PATH variable.
2. For Maven be sure to configure a server in the settings.xml file at the .m2 folder. This is needed as it contains the credentials to deploy the artifact to the Nexus maven repository.

Example of a settings.xml file:
```xml
<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
                      https://maven.apache.org/xsd/settings-1.0.0.xsd">
  <servers>
    <server>
      <id>{server_name}</id>
      <username>{nexus_user_login}</username>
      <password>{passwd}</password>
    </server>
  </servers>
</settings>
```

This server id the "--repositoryId" argument, so the mvn command will get this username and password to proceed with the deploy.

**HINT**
Never expose your password in plain text. For maven password encryption , follow [this guide](https://maven.apache.org/guides/mini/guide-encryption.html).


## Usage

```
nilegacy --help
```

The command syntax is:
```bash
nilegacy [repo] [options] directory_to_search_recursively
```

### Repo

In the repo argument you should use one of following options:
* maven
* npm
* nuget


### Options

General options (valid for maven, nuget and npm):
```
--repoUrl - repository URL
--version - show the current version
--help - display help
--output -change the output log file. Default is ./nilegacy-YYYYMMDDHHmmss.log
```

Options available only for maven:

```
--repositoryId - repository ID as configured in Maven settings.xml
--generatePom - if specified will force the pom generation, otherwise, will use the same name of the jar file for the pom
--groupId - group ID must be informed if generatePom is true
```

### Directory

The directory path to search for files. The execution will look for files with the following extensions / patterns:
* maven: *.jar
* nuget: *.nupkg
* npm: package.json and *.tgz
The CLI will recursively look for files except for those located in node_modules, considering NPM packages.

**Important**: Remember to check if the user have permission to access the directory.


### Importing Maven Artifacts
The CLI will look for files with the .jar extension and *.sources.jar and *.javadoc.jar are also considered. Behind the scenes it will execute the maven deploy plugin.
CLI example to upload maven artifacts to Nexus OSS:

```bash
nilegacy maven --repoUrl http://localhost:8081/repository/maven-hosted --repositoryId repo-maven /home/user/artifactory_
maven_backup
```

**Important**: Note the --repositoryId. This id must exists in the servers config at the settings.xml.

### Importing NPM Artifacts
This CLI will search recursively looking for package.json and *.tgz files, except those inside node_modules. 

First of all  you have to setup the user that will publish the package.

```bash
npm adduser --registry <repo_url>
```

Example:
```bash
npm adduser --registry http://localhost:8081/repository/npm-internal
```

Then run the nilegacy import command :

```bash
nilegacy npm --repoUrl http://localhost:8081/repository/npm-internal ./directory-with-npm-libs
```

**Limitations:**
NPM Issue [#10117](https://github.com/npm/npm/issues/10117)

### Importing NuGet Artifacts
TODO: To be done

## History
For the list of all changes see the [history log](CHANGELOG.md).


## License

Licensed under [MIT](LICENSE.md).

Copyright (c) 2017 Mauricio Gemelli Vigolo [@mauriciovigolo](https://twitter.com/mauriciovigolo)

CHANGELOG.md
# 0.2.0

First release, but still working in nuget upload. List of updates in this version
* Fixed issues #2 and #3.
* Improved documentation.
* Code refactors.
* Better support to execute the CLI on windows OS.

# 0.1.0-beta.5

Working on docs - install and usage.


# 0.1.0-beta.4

Improved the project documentation
Changed the bin command name to nilegacy

```bash
nilegacy -h
```

# 0.1.0-beta.1

This version updates the filename of /bin/nilegacy, regarding issue #1 .


# 0.1.0-beta.0

First release. Maven artifact upload is fully working.

This is a beta release. Need to work on the documentation and tests.

Next releases will deal with other processors - npm and nuget :smile:

test/nilegacy.spec.js
'use strict';

const assert = require('assert');
const spawn = require('child_process').spawnSync;


describe('nilegacy', function () {
  describe('Args Initialization', function () {
    it('No args: Should show usage options', function () {
      let result = spawn('node', ['./bin/nilegacy.js']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('USAGE:') > -1);
    });

    it('--help arg: Should show help', function () {
      let result = spawn('node', ['./bin/nilegacy.js', '--help']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('USAGE:') > -1);
    });

    it('-h arg: Should show help', function () {
      let result = spawn('node', ['./bin/nilegacy.js', '-h']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('USAGE:') > -1);
    });

    it('--version arg: Should show version', function () {
      let result = spawn('node', ['./bin/nilegacy.js', '--version']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Nexus Legacy Importer - v') > -1);
    });

    it('-v arg: Should show version', function () {
      let result = spawn('node', ['./bin/nilegacy.js', '-v']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Nexus Legacy Importer - v') > -1);
    });

    it('--examples arg: Should show examples', function () {
      let result = spawn('node', ['./bin/nilegacy.js', '--examples']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Nexus Legacy Importer - Examples') > -1);
    });

    it('Cannot accept --examples arg after another arg', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven', '--examples']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Invalid syntax') > -1);
    });

    it('Cannot accept --version arg after another arg', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven', '--version']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Invalid syntax') > -1);
    });

    it('Cannot accept --h arg after another arg', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven', '-h']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Invalid syntax') > -1);
    });

    it('Cannot accept --help arg after another arg', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven', '--help']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Invalid syntax') > -1);
    });

    it('Should validate repo types, accepting only maven, node and nuget', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'non-available-repo']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('nilegacy: Incorrect repository option') > -1);
    });

    it('Should not run if there is no repoUrl', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('the "--repoUrl" option is required') > -1);
    });

    it('Should return error if the directory containing the libraries doesn\'t exists', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven', '--repoUrl', 'http://localhost:8081/repo/', '/somedir/repo']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('Repository directory') > -1);
    });

    it('For Maven: Should throws error if it doesn\'t contain the repositoryId', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'maven', '--repoUrl', 'http://localhost:8081/repo/', '.']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('"--repositoryId" argument is required for maven repo type') > -1);
    });

    it('If !== Maven: Should not accept --repositoryId, --generatePom, --groupId', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'npm', '--repoUrl', 'http://localhost:8081/repo/', '--generatePom', 'true', '.']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('"--repositoryId", "--generatePom" and "--groupId" options are restricted to the maven repository') > -1);
    });

    it('For NuGet: throws error if it doesn\'t contain the apiKey', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'nuget', '--repoUrl', 'http://localhost:8081/repo/', '.']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('the "--apiKey" argument is required for nuget repo type') > -1);
    });

    it('If !== NuGet: Should not accept --apiKey', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'npm', '--apiKey', '<generated_key>', '--repoUrl', 'http://localhost:8081/repo/', '.']);

      let log = (result.stdout || result.stderr || '').toString();

      assert(log.indexOf('the "--apiKey" argument is restricted to the nuget repository') > -1);
    });
  });


  describe('Searching Files', function () {
    it('No files found, should exit', function () {
      let result = spawn('node', ['./bin/nilegacy.js', 'npm', '--repoUrl', 'http://localhost:8081/repo/', '/tmp/']);

      let log = (result.stdout || result.stderr || '').toString();
      console.log(log);

      assert(true);
    });
  });

});


index.js
var fm = require('file-matcher');
var spawn = require('child_process').spawnSync;

var matcher = new fm.FileMatcher();

var criteria = {
    path: '',
    recursiveSearch: true,
    fileFilter: {
        fileNamePattern: '**/*.jar'
    }
};


matcher.find(criteria)
    .then(function (files) {
        if (!files || files.length === 0) {
            console.log('No files found!');
            return;
        }

        for (var index = 0; index < files.length; index++) {
            var file = files[index];

            var args = [
                'deploy:deploy-file',
                '-DgeneratePOM=false',
                '-DrepositoryId=',
                '-Durl=',
                '-Dfile=' + file
            ];

            if (file.indexOf('-javadoc.jar') > -1) {
                var pomFile = file.substr(0, (file.length - 12)) + '.pom';

                args.push('-Dpackaging=javadoc');
                args.push('-DpomFile=' + pomFile);
            } else if (file.indexOf('-sources.jar') > -1) {
                var pomFile = file.substr(0, (file.length - 12)) + '.pom';

                args.push('-Dpackaging=java-source');
                args.push('-DpomFile=' + pomFile);
            } else if (file.indexOf('-client.jar') > -1) {
                var pomFile = file.substr(0, (file.length - 11)) + '.pom';

                args.push('-Dpackaging=client');
                args.push('-DpomFile=' + pomFile);
            } else {
                var pomFile = file.substr(0, (file.length - 4)) + '.pom';

                args.push('-DpomFile=' + pomFile);
            }

            var result = spawn('mvn', args);

            if (result.status !== 0) {
                console.log('File: ' + file, 'Output: ' + result.stdout);
            }

            mvn.stdout.on('data', (data) => {
                console.log(`stdout: ${data}`);
            });

            mvn.stderr.on('data', (data) => {
                console.log(`stderr: ${data}`);
            });

            mvn.on('close', (code) => {
                console.log(`child process exited with code ${code}`);
            });
        }
    })
    .catch(function (error) {
        console.log(error);
    });

lib/executable.js

function Executable(print) {

}

Executable.prototype.help = function() {
    var print = console.log;
    // repositoryId, generatePom, pomFile, groupId, version, artifactId
    print('USAGE: nilegacy [repo] [options] directory');
    print('');
    print('Repo:');
    print('%s\tspecify the type of repo (maven|node|nuget)', lPad('repo', 18));
    print('');
    print('Options:');
    print('%s\trepository URL', lPad('--repoUrl=', 18));
    print('%s\t[ONLY MAVEN] repository ID as configured in Maven\'s settings.xml', lPad('--repositoryId=', 18));
    print('%s\t[ONLY MAVEN] true to force the pom generation, otherwise, will use the same name of the jar file, just changing the extension.', lPad('--generatePom=', 18));
    print('%s\t[ONLY MAVEN] repository URL. If none, will use the pom as ', lPad('--groupId=', 18));
    print('%s\tshow the current version', lPad('--version', 18));
    print('%s\tdisplay this help', lPad('--help', 18));
    print('');
    print('Example:');
    print('nilegacy maven --repoUrl=http://localhost:8081/repository/maven-hosted');
}

Executable.prototype.version = function() {
    console.log('1.0.0');
}

function lPad(str, length) {
    if (str.length >= length) {
        return str;
    } else {
        return lPad(' ' + str, length);
    }
}

module.exports = Executable;

lib/processors/processor.js
/**
 * @license
 * Copyright (c) 2017 Mauricio Gemelli Vigolo. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/mauriciovigolo/nexus-ilegacy/blob/master/LICENSE
 */
var spawn = require('child_process').spawnSync;
var fs = require('fs');
var moment = require('moment');
var Progress = require('progress');
var MvnProcessor = require('./mvn.processor');
var NpmProcessor = require('./npm.processor');
var NugetProcessor = require('./nuget.processor');
var os = require('os');

var processor;
var progressBar;
var filesToExecute;
var outputLogFile;

function Processor(files, repoOptions) {
  filesToExecute = files;

  outputLogFile = repoOptions.outputLogFile;
  if (!outputLogFile) {
    outputLogFile = `nilegacy-${moment().format('YYYYMMDDHHmmss')}.log`;
  }

  try {
    switch (repoOptions.repo) {
      case 'maven':
        processor = new MvnProcessor(repoOptions.repoUrl, repoOptions.repositoryId, repoOptions.generatePom, repoOptions.groupId);
        break;
      case 'npm':
        processor = new NpmProcessor(repoOptions.repoUrl);
        break;
      case 'nuget':
        processor = new NugetProcessor(repoOptions.repoUrl, repoOptions.apiKey);
        break;
    }
  } catch (error) {
    throw new Error(`The ${repoOptions.repo} processor could not be initialized`);
  }
}

Processor.prototype.importToNexus = function () {
  initializeProgressBar(filesToExecute.length);

  var errors = 0;
  var success = 0;

  var command = processor.getConfigValue('command');
  command = addWindowsSuffix(command);

  filesToExecute.forEach(file => {
    var msg;
    var args = processor.prepareSpawn(file);
    var result = spawn(command, args);

    if (result.status !== 0) {
      var osMsg = result.stderr ? result.stderr.toString() : undefined;
      if (!osMsg) {
        osMsg = result.stdout ? result.stdout.toString() : ' - ';
      }
      msg = `ERROR ${moment().format('YYYY-MM-DD HH:mm:ss')}: Repository uploading error for file: ${file}.\nDetails:\n${osMsg}`;
      errors += 1;
    } else {
      msg = `INFO ${moment().format('YYYY-MM-DD HH:mm:ss')}: File ${file} was successfully uploaded`;
      success += 1;
    }

    fs.appendFile(outputLogFile, msg, { encoding: 'utf8', flag: 'a' })

    progressBar.tick({
      success: success,
      errors: errors
    });
  });
}

Processor.prototype.isConfigured = function () {
  var command = processor.getConfigValue('command');
  command = addWindowsSuffix(command);
  var args = processor.getConfigValue('args');
  var result = spawn(command, [args.help]);

  return !result.error;
}

function addWindowsSuffix(command) {
  var commandSuffix = os.type().toLocaleLowerCase().indexOf('windows') > -1 ? '.cmd' : '';
  return command + commandSuffix;
}

function initializeProgressBar(total) {
  progressBar = new Progress(`  Nexus Import Progress [:bar] :percent | Eta: :eta s | Totals: success: :success | errors: :errors |`, {
    complete: '=',
    incomplete: '.',
    width: 20,
    total: total
  });
}


module.exports = Processor;

lib/processors/mvn.processor.js
/**
 * @license
 * Copyright (c) 2017 Mauricio Gemelli Vigolo. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/mauriciovigolo/nexus-ilegacy/blob/master/LICENSE
 */
var config = {
    command: 'mvn',
    description: 'Apache Maven',
    args: {
        help: '-h'
    }        
};

/**
 *
 *
 * @param {string} repoUrl
 * @param {string} repositoryId
 * @param {boolean} generatePom
 * @param {string} groupId
 */
function MvnProcessor(repoUrl, repositoryId, generatePom, groupId) {
    config.repoUrl = repoUrl;
    config.repositoryId = repositoryId;
    config.generatePom = generatePom;
    config.groupId = groupId;

    if (!config.repoUrl) {
        throw new Error('Repository URL can\'t be null or undefined');
    }

    if (!config.repositoryId) {
        throw new Error('repositoryId can\'t be null or undefined');
    }
}

MvnProcessor.prototype.getConfigValue = function (key) {
    return config[key];
}

MvnProcessor.prototype.prepareSpawn = function (file) {
    var args = [
        `deploy:deploy-file`,
        `-Dfile=${file}`,
        `-Durl=${config.repoUrl}`,
        `-DrepositoryId=${config.repositoryId}`
    ];

    var pomFile;
    // Package definition
    if (file.indexOf('-javadoc.jar') > -1) {
        pomFile = file.substr(0, (file.length - 12)) + '.pom';
        args.push('-Dpackaging=javadoc');
    } else if (file.indexOf('-sources.jar') > -1) {
        pomFile = file.substr(0, (file.length - 12)) + '.pom';
        args.push('-Dpackaging=java-source');
    } else {
        pomFile = file.substr(0, (file.length - 4)) + '.pom';
    }

    if (!config.generatePom) {
        args.push(
            `-DgeneratePOM=false`,
            `-DpomFile=${pomFile}`
        );
    } else if (config.groupId) {
        args.push(
            `-DgroupId=${config.groupId}`
        );
    }

    return args;
}

module.exports = MvnProcessor;

lib/processors/npm.processor.js
/**
 * @license
 * Copyright (c) 2017 Mauricio Gemelli Vigolo. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/mauriciovigolo/nexus-ilegacy/blob/master/LICENSE
 */
var path = require('path');

var config = {
    command: 'npm',
    description: 'Node Package Manager',
    args: {
        help: 'help'
    }    
};

function NpmProcessor(repoUrl) {
    config.repoUrl = repoUrl;

    if (!config.repoUrl) {
        throw new Error('Repository URL can\'t be null or undefined');
    }
}

NpmProcessor.prototype.getConfigValue = function (key) {
    return config[key];
}

NpmProcessor.prototype.prepareSpawn = function (file) {
    if (!file) {
        throw new Error('Filename can\'t be null or undefined');
    }

    var dir = path.dirname(file);

    var args = [
      'publish',
      dir,
      `--"@scope:registry=${config.repoUrl}`
    ];

    return args;
}

module.exports = NpmProcessor;

lib/processors/nuget.processor.js
/**
 * @license
 * Copyright (c) 2017 Mauricio Gemelli Vigolo. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/mauriciovigolo/nexus-ilegacy/blob/master/LICENSE
 */
var config = {
  command: 'nuget',
  description: 'Microsoft Package Manager',
  args: {
    help: 'help'
  }
};

function NugetProcessor(repoUrl, apiKey) {
  config.repoUrl = repoUrl;
  config.apiKey = apiKey;

  if (!config.repoUrl || !config.apiKey) {
    throw new Error('Repository URL and api key can\'t be null or undefined');
  }
}

NugetProcessor.prototype.getConfigValue = function (key) {
  return config[key];
}

NugetProcessor.prototype.prepareSpawn = function (file) {
  if (!file) {
    throw new Error('Filename can\'t be null or undefined');
  }

  var args = [
    'push',
    file,
    config.apiKey,
    '-Source',
    config.repoUrl
  ];

  return args;
}

module.exports = NugetProcessor;

lib/nilegacy.js
/**
 * @license
 * Copyright (c) 2017 Mauricio Gemelli Vigolo. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/mauriciovigolo/nexus-ilegacy/blob/master/LICENSE
 */
var fm = require('file-matcher');
var fs = require('fs');
var spawn = require('child_process').spawnSync;
var Processor = require('./processors/processor');
var chalk = require('chalk');
var package = require('../package.json');

var print;
var exit;
var processor;
var existsSync;
var accessSync;
var repoDir;
var repoOptions;

var info;
var warn;

function NiLegacy(args) {
  print = console.log;
  exit = process.exit;
  existsSync = fs.existsSync;
  accessSync = fs.accessSync;

  info = chalk.yellow;
  warn = chalk.red;

  try {
    loadProcessorOptions(args);
    searchFiles()
      .then(files => {
        processor = new Processor(files, repoOptions);

        if (!processor.isConfigured()) {
          printError('%s was not found or configured. Checkout nexus-ilegacy docs for more info - https://github.com/mauriciovigolo/nexus-ilegacy/wiki', null, repoOptions.repo);
        }

        processor.importToNexus();
      })
      .catch(error => {
        printError(error);
      });
  } catch (error) {
    printError(error);
  }
}

function loadProcessorOptions(args) {
  if (!args || args.length === 0 || args[0] === '-h' || args[0] === '--help') {
    help();
  }

  if (args[0] === '-v' || args[0] === '--version') {
    version();
  }

  if (args[0] === '--examples') {
    examples();
  }

  var repo = args.shift();
  if (repo !== 'maven' && repo !== 'npm' && repo !== 'nuget') {
    printError('Incorrect repository option: %s', 'Valid repo keys: maven | npm | nuget', repo);
  }

  var repoUrl;
  var repositoryId;
  var generatePom;
  var groupId;
  var apiKey;
  var outputLogFile;

  while (args.length) {
    var arg = args.shift();

    switch (arg) {
      case '--repoUrl':
        repoUrl = shiftArg(args);
        break;
      case '--repositoryId':
        repositoryId = shiftArg(args);
        break;
      case '--generatePom':
        generatePom = true;
        break;
      case '--apiKey':
        apiKey = shiftArg(args);
        break;
      case '--groupId':
        groupId = shiftArg(args);
        break;
      case '--examples':
        printError('Invalid syntax. For examples type: nilegacy --examples');
        break;
      case '-h':
        printError('Invalid syntax. For help type: nilegacy --help');
        break;
      case '--help':
        printError('Invalid syntax. For help type: nilegacy --help');
        break;
      case '--version':
        printError('Invalid syntax. For version info type: nilegacy --version');
        break;
      case '--output':
        outputLogFile = shiftArg(args);
        if (!accessSync(outputLogFile)) {
          printError('User doesn\'t have permission to access %s', outputLogFile);
        }
        break;
      default:
        repoDir = arg;
        break;
    }
  }

  if (!repoUrl) {
    printError('the "--repoUrl" option is required');
  }

  if (!existsSync(repoDir)) {
    printError('Repository directory "%s" doesn\'t exist!', null, repoDir);
  }

  repoOptions = {
    repo: repo,
    repoUrl: repoUrl,
    outputLogFile: outputLogFile,
    groupId: groupId,
    generatePom: generatePom,
    repositoryId: repositoryId,
    apiKey: apiKey
  }

  if ((repo !== 'maven') && (repositoryId || generatePom || groupId)) {
    printError('"--repositoryId", "--generatePom" and "--groupId" options are restricted to the maven repository');
  } else if ((repo === 'maven') && !repositoryId) {
    printError('the "--repositoryId" argument is required for maven repo type');
  }

  if ((repo !== 'nuget') && (apiKey)) {
    printError('the "--apiKey" argument is restricted to the nuget repository');
  } else if ((repo === 'nuget') && !apiKey) {
    printError('the "--apiKey" argument is required for nuget repo type');
  }
}

function searchFiles() {
  var matcher = new fm.FileMatcher();

  var fileNamePattern;
  if (repoOptions.repo === 'maven') {
    fileNamePattern = '**/**.jar';
  } else if (repoOptions.repo === 'npm') {
    fileNamePattern = ['**/package.json', '**/**.tgz', '!node_modules'];
  } else if (repoOptions.repo === 'nuget') {
    fileNamePattern = ['**/**.nupkg'];
  }

  var criteria = {
    path: repoDir,
    recursiveSearch: true,
    fileFilter: {
      fileNamePattern: fileNamePattern
    }
  };

  return new Promise((resolve, reject) => {
    matcher.find(criteria)
      .then(function (files) {
        if (!files || !files.length) {
          reject('No files found!');
        }

        resolve(files);
      })
      .catch(function (error) {
        reject('An error happened during file search');
      });
  });
}

function help() {
  print('Nexus Legacy Importer - Help');
  print('%s: nilegacy [repo] [options] directory', warn('USAGE'));
  print('');
  print('Repo:');
  print('%s\tspecify the type of repo (maven|npm|nuget)', info(lPad('repo', 18)));
  print('');
  print('Options:');
  print('%s\trepository URL', info(lPad('--repoUrl', 18)));
  print('%s\tshow the current version', info(lPad('--version', 18)));
  print('%s\tdisplay this help', info(lPad('--help', 18)));
  print('%s\tchange the output log file. Default is ./nilegacy-YYYYMMDDHHmmss.log', info(lPad('--output', 18)));
  print('%s\tshows examples for maven, nuget and npm', info(lPad('--examples', 18)));
  print();
  print('Options valid ONLY for %s repo', warn('NuGet'));
  print('%s\tAPI Key for the nuget repo access', info(lPad('--apiKey', 18)));
  print();
  print('Options valid ONLY for %s repo', warn('Maven'));
  print('%s\trepository ID as configured in Maven\'s settings.xml', info(lPad('--repositoryId', 18)));
  print('%s\tif specified will force the pom generation, otherwise, will use the same name of the jar file for the pom', info(lPad('--generatePom', 18)));
  print('%s\tgroup ID must be informed if generatePom is true', info(lPad('--groupId', 18)));

  exit(0);
}

function examples() {
  print('Nexus Legacy Importer - Examples');
  print('');
  print(info('Maven'));
  print('nilegacy maven --repoUrl http://localhost:8081/repository/maven-hosted/ --repositoryId maven-localhost /home/user/maven_repo');
  print('%s Looks for *.jar files recursively and publish them in the specified repo', warn('what it does: '));
  print('');
  print(info('NPM'));
  print('nilegacy npm --repoUrl http://localhost:8081/repository/node-hosted/ /home/user/node_repo');
  print('%s Looks for package.json files recursively, not including node_modules files, and publish them in the specified repo', warn('what it does: '));
  print('');
  print(info('NuGet'));
  print('nilegacy nuget --repoUrl http://localhost:8081/repository/nuget-hosted/ --apiKey 47be3377-c434-4c29-8576-af7f6993a54b /home/user/nuget_repo');
  print('%s Looks for *.nupkg files recursively and publish them in the specified repo', warn('what it does: '));
  exit(0);
}

function version() {
  print('Nexus Legacy Importer - v%s', package.version);
  print('License under MIT. Copyright (c) 2017 Mauricio Gemelli Vigolo');

  exit(0);
}

function shiftArg(args) {
  if (!args || args.length === 0 || args[0].indexOf('--') > -1) {
    return;
  }
  return args.shift();
}

function lPad(str, length) {
  if (str.length >= length) {
    return str;
  } else {
    return lPad(' ' + str, length);
  }
}

function printError(msg, details, args) {
  print('nilegacy: ' + msg, args || '');
  if (details) {
    print(details);
  }
  print();
  print('Try "nilegacy --help" for help and more options');
  exit(1);
}

module.exports = NiLegacy;
